File: python/src/gwp_py/transaction.py
"""Transaction management."""

from __future__ import annotations

from typing import Any

from gwp_py._convert import value_to_proto
from gwp_py._generated import gql_service_pb2 as gql_pb2
from gwp_py._generated import gql_types_pb2 as types_pb2
from gwp_py.errors import GqlStatusError, TransactionError
from gwp_py.result import ResultCursor
from gwp_py.status import is_exception


class Transaction:
    """An explicit transaction within a session."""

    def __init__(
        self,
        session_id: str,
        transaction_id: str,
        gql_stub: Any,
    ):
        self._session_id = session_id
        self._transaction_id = transaction_id
        self._gql_stub = gql_stub
        self._committed = False
        self._rolled_back = False

    @classmethod
    async def begin(
        cls,
        session_id: str,
        gql_stub: Any,
        mode: int,
    ) -> Transaction:
        """Begin a new transaction."""
        resp = await gql_stub.BeginTransaction(
            gql_pb2.BeginRequest(session_id=session_id, mode=mode)
        )

        if resp.status and is_exception(resp.status.code):
            raise GqlStatusError(resp.status.code, resp.status.message)

        if not resp.transaction_id:
            raise TransactionError("server returned empty transaction ID")

        return cls(session_id, resp.transaction_id, gql_stub)

    @property
    def transaction_id(self) -> str:
        """The transaction identifier."""
        return self._transaction_id

    async def execute(
        self,
        statement: str,
        parameters: dict[str, Any] | None = None,
    ) -> ResultCursor:
        """Execute a statement within this transaction."""
        proto_params = {}
        if parameters:
            for k, v in parameters.items():
                proto_params[k] = value_to_proto(v, types_pb2)

        stream = self._gql_stub.Execute(
            gql_pb2.ExecuteRequest(
                session_id=self._session_id,
                statement=statement,
                parameters=proto_params,
                transaction_id=self._transaction_id,
            )
        )

        return ResultCursor(stream)

    async def commit(self) -> None:
        """Commit the transaction."""
        resp = await self._gql_stub.Commit(
            gql_pb2.CommitRequest(
                session_id=self._session_id,
                transaction_id=self._transaction_id,
            )
        )
        self._committed = True

        if resp.status and is_exception(resp.status.code):
            raise GqlStatusError(resp.status.code, resp.status.message)

    async def rollback(self) -> None:
        """Roll back the transaction."""
        if self._committed or self._rolled_back:
            return

        resp = await self._gql_stub.Rollback(
            gql_pb2.RollbackRequest(
                session_id=self._session_id,
                transaction_id=self._transaction_id,
            )
        )
        self._rolled_back = True

        if resp.status and is_exception(resp.status.code):
            raise GqlStatusError(resp.status.code, resp.status.message)

    async def __aenter__(self) -> Transaction:
        return self

    async def __aexit__(self, exc_type: type | None, *args: object) -> None:
        if exc_type is not None:
            await self.rollback()
        elif not self._committed:
            await self.commit()

File: java/src/main/java/dev/grafeo/gwp/Transaction.java
package dev.grafeo.gwp;

import dev.grafeo.gwp.errors.GqlStatusException;
import dev.grafeo.gwp.errors.TransactionException;
import dev.grafeo.gwp.internal.ValueConverter;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * An explicit transaction within a session.
 *
 * <p>Implements {@link AutoCloseable} for use with try-with-resources. If the
 * transaction has not been committed when {@link #close()} is called, it will
 * be automatically rolled back.</p>
 *
 * <pre>{@code
 * try (Transaction tx = session.beginTransaction()) {
 *     tx.execute("INSERT (:Person {name: 'Alice'})");
 *     tx.commit();
 * }
 * // If commit() was not called, rollback happens automatically
 * }</pre>
 */
public class Transaction implements AutoCloseable {

    private final String sessionId;
    private final String transactionId;
    private final gql.GqlServiceGrpc.GqlServiceBlockingStub gqlStub;
    private final gql.SessionServiceGrpc.SessionServiceBlockingStub sessionStub;
    private boolean committed = false;
    private boolean rolledBack = false;

    Transaction(
            String sessionId,
            String transactionId,
            gql.GqlServiceGrpc.GqlServiceBlockingStub gqlStub,
            gql.SessionServiceGrpc.SessionServiceBlockingStub sessionStub) {
        this.sessionId = sessionId;
        this.transactionId = transactionId;
        this.gqlStub = gqlStub;
        this.sessionStub = sessionStub;
    }

    /**
     * Begin a new transaction.
     *
     * @param sessionId   the session ID
     * @param gqlStub     the GQL service stub
     * @param sessionStub the session service stub
     * @param readOnly    true for a read-only transaction
     * @return the new transaction
     * @throws GqlStatusException  if the server returns an exception status
     * @throws TransactionException if the server returns an empty transaction ID
     */
    static Transaction begin(
            String sessionId,
            gql.GqlServiceGrpc.GqlServiceBlockingStub gqlStub,
            gql.SessionServiceGrpc.SessionServiceBlockingStub sessionStub,
            boolean readOnly) {

        gql.GqlServiceOuterClass.TransactionMode mode = readOnly
                ? gql.GqlServiceOuterClass.TransactionMode.READ_ONLY
                : gql.GqlServiceOuterClass.TransactionMode.READ_WRITE;

        gql.GqlServiceOuterClass.BeginResponse resp = gqlStub.beginTransaction(
                gql.GqlServiceOuterClass.BeginRequest.newBuilder()
                        .setSessionId(sessionId)
                        .setMode(mode)
                        .build());

        if (resp.hasStatus() && GqlStatus.isException(resp.getStatus().getCode())) {
            throw new GqlStatusException(
                    resp.getStatus().getCode(),
                    resp.getStatus().getMessage());
        }

        if (resp.getTransactionId().isEmpty()) {
            throw new TransactionException("server returned empty transaction ID");
        }

        return new Transaction(sessionId, resp.getTransactionId(), gqlStub, sessionStub);
    }

    /** The transaction identifier. */
    public String transactionId() {
        return transactionId;
    }

    /**
     * Execute a GQL statement within this transaction.
     *
     * @param statement the GQL statement to execute
     * @return a cursor over the results
     */
    public ResultCursor execute(String statement) {
        return execute(statement, null);
    }

    /**
     * Execute a GQL statement within this transaction with parameters.
     *
     * @param statement  the GQL statement to execute
     * @param parameters named parameters (may be null)
     * @return a cursor over the results
     */
    public ResultCursor execute(String statement, Map<String, Object> parameters) {
        gql.GqlServiceOuterClass.ExecuteRequest.Builder reqBuilder =
                gql.GqlServiceOuterClass.ExecuteRequest.newBuilder()
                        .setSessionId(sessionId)
                        .setStatement(statement)
                        .setTransactionId(transactionId);

        if (parameters != null && !parameters.isEmpty()) {
            Map<String, gql.GqlTypes.Value> protoParams = new HashMap<>(parameters.size());
            for (Map.Entry<String, Object> entry : parameters.entrySet()) {
                protoParams.put(entry.getKey(), ValueConverter.toProto(entry.getValue()));
            }
            reqBuilder.putAllParameters(protoParams);
        }

        Iterator<gql.GqlServiceOuterClass.ExecuteResponse> stream =
                gqlStub.execute(reqBuilder.build());

        return new ResultCursor(stream);
    }

    /**
     * Commit the transaction.
     *
     * @throws GqlStatusException if the server returns an exception status
     */
    public void commit() {
        gql.GqlServiceOuterClass.CommitResponse resp = gqlStub.commit(
                gql.GqlServiceOuterClass.CommitRequest.newBuilder()
                        .setSessionId(sessionId)
                        .setTransactionId(transactionId)
                        .build());

        committed = true;

        if (resp.hasStatus() && GqlStatus.isException(resp.getStatus().getCode())) {
            throw new GqlStatusException(
                    resp.getStatus().getCode(),
                    resp.getStatus().getMessage());
        }
    }

    /**
     * Roll back the transaction.
     *
     * <p>This is a no-op if the transaction has already been committed or
     * rolled back.</p>
     *
     * @throws GqlStatusException if the server returns an exception status
     */
    public void rollback() {
        if (committed || rolledBack) {
            return;
        }

        gql.GqlServiceOuterClass.RollbackResponse resp = gqlStub.rollback(
                gql.GqlServiceOuterClass.RollbackRequest.newBuilder()
                        .setSessionId(sessionId)
                        .setTransactionId(transactionId)
                        .build());

        rolledBack = true;

        if (resp.hasStatus() && GqlStatus.isException(resp.getStatus().getCode())) {
            throw new GqlStatusException(
                    resp.getStatus().getCode(),
                    resp.getStatus().getMessage());
        }
    }

    /**
     * Close the transaction. If the transaction has not been committed,
     * it will be rolled back automatically.
     */
    @Override
    public void close() {
        if (!committed && !rolledBack) {
            try {
                rollback();
            } catch (Exception e) {
                // Suppress rollback errors during close
            }
        }
    }
}
